File: blackbox/a2a-lifecycle-terminal-v0.1/runner.mjs
#!/usr/bin/env node
import { readFile } from "node:fs/promises";
import { fileURLToPath } from "node:url";
import { dirname, join } from "node:path";
import {
  PROFILE,
  canonical,
  evaluateLifecycleTrace,
  parseStrictJson,
  projectReceipt,
  sha256
} from "./evaluator.mjs";

const root = dirname(fileURLToPath(import.meta.url));

async function load() {
  const contractBytes = await readFile(join(root, "contract.json"));
  const corpusBytes = await readFile(join(root, "corpus/v0.1/cases.json"));
  const contract = parseStrictJson(contractBytes.toString("utf8"));
  const corpus = parseStrictJson(corpusBytes.toString("utf8"));
  if (contract.profile !== PROFILE || corpus.profile !== PROFILE) throw new Error("profile mismatch");
  if (canonical(corpus.mandatory_case_ids) !== canonical(corpus.cases.map((item) => item.id))) {
    throw new Error("mandatory_case_ids must exactly match cases in order");
  }
  if (canonical(corpus.supplemental_case_ids) !== canonical(corpus.supplemental_cases.map((item) => item.id))) {
    throw new Error("supplemental_case_ids must exactly match supplemental_cases in order");
  }
  return { contractBytes, corpusBytes, corpus };
}

function expectedEvents(trace, stepCodes) {
  if (stepCodes.length === 0) return [];
  let workId = null;
  let attemptNumber = 0;
  let ownerEpoch = 0;
  let maxAttempts = 0;
  const events = [];
  const push = (kind, at) => {
    const seq = events.length + 1;
    events.push({
      id: `${workId}:event:${seq}`,
      seq,
      kind,
      attempt_id: `${workId}:attempt:${attemptNumber}`,
      owner_epoch: ownerEpoch,
      occurred_at: at
    });
  };
  trace.commands.forEach((command, index) => {
    if (stepCodes[index] !== "OK") return;
    switch (command.op) {
      case "create":
        workId = command.work_id;
        attemptNumber = 1;
        ownerEpoch = 0;
        maxAttempts = command.max_attempts;
        push("attempt_created", command.at);
        break;
      case "acquire":
        ownerEpoch += 1;
        push("lease_acquired", command.at);
        break;
      case "renew":
        push("lease_acquired", command.at);
        break;
      case "expire":
        push("lease_expired", command.at);
        if (attemptNumber < maxAttempts) {
          attemptNumber += 1;
          ownerEpoch += 1;
          push("attempt_retried", command.at);
        } else {
          push("attempt_failed", command.at);
        }
        break;
      case "settle":
        push(command.outcome === "success" ? "attempt_succeeded" : "attempt_failed", command.at);
        break;
      case "settle_with_retry":
        push("attempt_failed", command.at);
        if (attemptNumber < maxAttempts) {
          attemptNumber += 1;
          ownerEpoch += 1;
          push("attempt_retried", command.at);
        }
        break;
      case "cancel":
        push("attempt_cancelled", command.at);
        break;
    }
  });
  return events;
}

function executeCases(corpus) {
  return [...corpus.cases, ...corpus.supplemental_cases].map((item) => {
    const receipt = evaluateLifecycleTrace(item.trace);
    const actual = projectReceipt(receipt);
    const metadataPassed = receipt.final === null
      ? expectedEvents(item.trace, item.expected.step_codes).length === 0
      : canonical(receipt.final.events) === canonical(expectedEvents(item.trace, item.expected.step_codes));
    return {
      id: item.id,
      passed: canonical(actual) === canonical(item.expected) && metadataPassed,
      metadata_passed: metadataPassed,
      expected_sha256: sha256(item.expected),
      actual_sha256: sha256(actual),
      receipt
    };
  });
}

async function main() {
  const { contractBytes, corpusBytes, corpus } = await load();
  const results = executeCases(corpus);
  if (process.argv.includes("--emit-cases")) {
    if (results.some((item) => !item.passed)) throw new Error("cannot emit a failing corpus");
    process.stdout.write(`${canonical(results.map((item) => item.receipt))}\n`);
    return;
  }
  if (process.argv.includes("--self-test")) {
    if (results.some((item) => !item.passed)) throw new Error("baseline corpus failed");
    let mutationDetections = 0;
    for (const item of [...corpus.cases, ...corpus.supplemental_cases]) {
      const mutated = JSON.parse(JSON.stringify(item.expected));
      mutated.step_codes.push("__MUTATED__");
      if (canonical(projectReceipt(evaluateLifecycleTrace(item.trace))) !== canonical(mutated)) mutationDetections += 1;
    }
    const controls = [
      () => parseStrictJson("{\"a\":1,\"a\":2}"),
      () => parseStrictJson("{\"a\":1.5}"),
      () => parseStrictJson("{\"a\":-0}"),
      () => parseStrictJson("{\"a\":9007199254740992}"),
      () => parseStrictJson("{\"a\":\"\\ud800\"}"),
      () => parseStrictJson("\u00a0{\"a\":1}"),
      () => parseStrictJson("\ufeff{\"a\":1}")
    ];
    let parserDetections = 0;
    for (const control of controls) {
      try {
        control();
      } catch {
        parserDetections += 1;
      }
    }
    const canonicalControls = [
      canonical(parseStrictJson("{\"__proto__\":1,\"10\":2,\"2\":3}")) === "{\"10\":2,\"2\":3,\"__proto__\":1}",
      canonical(parseStrictJson("{\"\\ud800\\udc00\":5,\"\\ue000\":4}")) === "{\"\":4,\"𐀀\":5}"
    ].filter(Boolean).length;
    if (mutationDetections !== results.length || parserDetections !== controls.length || canonicalControls !== 2) {
      throw new Error("self-test control escaped detection");
    }
    process.stdout.write(`${canonical({
      profile: PROFILE,
      implementation: "javascript",
      corpus_cases: results.length,
      mandatory_cases: corpus.cases.length,
      supplemental_cases: corpus.supplemental_cases.length,
      baseline_passed: results.length,
      mutation_controls_passed: mutationDetections,
      parser_controls_passed: parserDetections,
      canonical_controls_passed: canonicalControls
    })}\n`);
    return;
  }
  const failed = results.filter((item) => !item.passed);
  const transcript = results.map(({ id, passed, expected_sha256, actual_sha256, receipt }) => ({
    id, passed, expected_sha256, actual_sha256, receipt_sha256: receipt.receipt_sha256
  }));
  const report = {
    profile: PROFILE,
    implementation: "javascript",
    contract_sha256: sha256(contractBytes.toString("utf8")),
    corpus_sha256: sha256(corpusBytes.toString("utf8")),
    case_count: results.length,
    passed: results.length - failed.length,
    failed: failed.length,
    failed_case_ids: failed.map((item) => item.id),
    transcript_sha256: sha256(transcript)
  };
  process.stdout.write(`${canonical(report)}\n`);
  if (failed.length > 0) process.exitCode = 1;
}

main().catch((error) => {
  process.stderr.write(`${error.stack ?? error}\n`);
  process.exitCode = 1;
});

File: blackbox/a2a-lifecycle-terminal-v0.1/python/runner.py
#!/usr/bin/env python3
import json
import pathlib
import sys

from evaluator import PROFILE, canonical, evaluate_lifecycle_trace, parse_strict_json, project_receipt, sha256

# Byte-differential contract: stdout is UTF-8 with "\n" newlines on every platform. The five
# runners patched in #106 proved this line on real windows-2022 CI; the unpatched ones failed
# there (#108 first run) -- Windows text-mode stdout emits \r\n from print(), bytes the
# JavaScript side of the differential never emits, and cp1252 cannot encode all corpus content.
sys.stdout.reconfigure(encoding="utf-8", newline="\n")


ROOT = pathlib.Path(__file__).resolve().parent.parent


def load():
    contract_text = (ROOT / "contract.json").read_text(encoding="utf-8")
    corpus_text = (ROOT / "corpus/v0.1/cases.json").read_text(encoding="utf-8")
    contract = parse_strict_json(contract_text)
    corpus = parse_strict_json(corpus_text)
    if contract["profile"] != PROFILE or corpus["profile"] != PROFILE:
        raise ValueError("profile mismatch")
    if canonical(corpus["mandatory_case_ids"]) != canonical([item["id"] for item in corpus["cases"]]):
        raise ValueError("mandatory_case_ids must exactly match cases in order")
    if canonical(corpus["supplemental_case_ids"]) != canonical([item["id"] for item in corpus["supplemental_cases"]]):
        raise ValueError("supplemental_case_ids must exactly match supplemental_cases in order")
    return contract_text, corpus_text, corpus


def expected_events(trace, step_codes):
    if not step_codes:
        return []
    work_id = None
    attempt_number = 0
    owner_epoch = 0
    max_attempts = 0
    events = []

    def push(kind, at):
        seq = len(events) + 1
        events.append({
            "id": f"{work_id}:event:{seq}",
            "seq": seq,
            "kind": kind,
            "attempt_id": f"{work_id}:attempt:{attempt_number}",
            "owner_epoch": owner_epoch,
            "occurred_at": at,
        })

    for index, command in enumerate(trace.get("commands", [])):
        if step_codes[index] != "OK":
            continue
        op = command["op"]
        if op == "create":
            work_id = command["work_id"]
            attempt_number = 1
            owner_epoch = 0
            max_attempts = command["max_attempts"]
            push("attempt_created", command["at"])
        elif op == "acquire":
            owner_epoch += 1
            push("lease_acquired", command["at"])
        elif op == "renew":
            push("lease_acquired", command["at"])
        elif op == "expire":
            push("lease_expired", command["at"])
            if attempt_number < max_attempts:
                attempt_number += 1
                owner_epoch += 1
                push("attempt_retried", command["at"])
            else:
                push("attempt_failed", command["at"])
        elif op == "settle":
            push("attempt_succeeded" if command["outcome"] == "success" else "attempt_failed", command["at"])
        elif op == "settle_with_retry":
            push("attempt_failed", command["at"])
            if attempt_number < max_attempts:
                attempt_number += 1
                owner_epoch += 1
                push("attempt_retried", command["at"])
        elif op == "cancel":
            push("attempt_cancelled", command["at"])
    return events


def execute_cases(corpus):
    results = []
    for item in corpus["cases"] + corpus["supplemental_cases"]:
        receipt = evaluate_lifecycle_trace(item["trace"])
        actual = project_receipt(receipt)
        expected_metadata = expected_events(item["trace"], item["expected"]["step_codes"])
        metadata_passed = (not expected_metadata) if receipt["final"] is None else canonical(receipt["final"]["events"]) == canonical(expected_metadata)
        results.append({
            "id": item["id"],
            "passed": canonical(actual) == canonical(item["expected"]) and metadata_passed,
            "metadata_passed": metadata_passed,
            "expected_sha256": sha256(item["expected"]),
            "actual_sha256": sha256(actual),
            "receipt": receipt,
        })
    return results


def main():
    contract_text, corpus_text, corpus = load()
    results = execute_cases(corpus)
    if "--emit-cases" in sys.argv:
        if any(not item["passed"] for item in results):
            raise ValueError("cannot emit a failing corpus")
        print(canonical([item["receipt"] for item in results]))
        return 0
    if "--self-test" in sys.argv:
        if any(not item["passed"] for item in results):
            raise ValueError("baseline corpus failed")
        mutation_detections = 0
        for item in corpus["cases"] + corpus["supplemental_cases"]:
            mutated = json.loads(json.dumps(item["expected"]))
            mutated["step_codes"].append("__MUTATED__")
            if canonical(project_receipt(evaluate_lifecycle_trace(item["trace"]))) != canonical(mutated):
                mutation_detections += 1
        controls = ['{"a":1,"a":2}', '{"a":1.5}', '{"a":-0}', '{"a":9007199254740992}', '{"a":"\\ud800"}', '\u00a0{"a":1}', '\ufeff{"a":1}']
        parser_detections = 0
        for control in controls:
            try:
                parse_strict_json(control)
            except (ValueError, UnicodeError):
                parser_detections += 1
        canonical_controls = sum([
            canonical(parse_strict_json('{"__proto__":1,"10":2,"2":3}')) == '{"10":2,"2":3,"__proto__":1}',
            canonical(parse_strict_json('{"\\ud800\\udc00":5,"\\ue000":4}')) == '{"\ue000":4,"\U00010000":5}',
        ])
        if mutation_detections != len(results) or parser_detections != len(controls) or canonical_controls != 2:
            raise ValueError("self-test control escaped detection")
        print(canonical({
            "profile": PROFILE,
            "implementation": "python",
            "corpus_cases": len(results),
            "mandatory_cases": len(corpus["cases"]),
            "supplemental_cases": len(corpus["supplemental_cases"]),
            "baseline_passed": len(results),
            "mutation_controls_passed": mutation_detections,
            "parser_controls_passed": parser_detections,
            "canonical_controls_passed": canonical_controls,
        }))
        return 0
    failed = [item for item in results if not item["passed"]]
    transcript = [{
        "id": item["id"],
        "passed": item["passed"],
        "expected_sha256": item["expected_sha256"],
        "actual_sha256": item["actual_sha256"],
        "receipt_sha256": item["receipt"]["receipt_sha256"],
    } for item in results]
    print(canonical({
        "profile": PROFILE,
        "implementation": "python",
        "contract_sha256": sha256(contract_text),
        "corpus_sha256": sha256(corpus_text),
        "case_count": len(results),
        "passed": len(results) - len(failed),
        "failed": len(failed),
        "failed_case_ids": [item["id"] for item in failed],
        "transcript_sha256": sha256(transcript),
    }))
    return 1 if failed else 0


if __name__ == "__main__":
    try:
        raise SystemExit(main())
    except Exception as error:
        print(f"{type(error).__name__}: {error}", file=sys.stderr)
        raise SystemExit(1)
